File: resources/android/StatusLabelRenderer.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import com.nativephp.mobile.ui.nativerender.NativeUIBridge
import com.nativephp.mobile.ui.nativerender.NativeUINode
import com.nativephp.mobile.ui.nativerender.NativeUITree
import com.nativephp.plugins.native_ui.NativeUITheme

data class StatusLabelRendererConfiguration(
    val label: String,
    val tone: StatusLabelTone,
    val accessibilityLabel: String,
    val accessibilityHint: String,
) {
    constructor(node: NativeUINode) : this(
        label = node.props.getString("label"),
        tone = StatusLabelTone.fromWire(node.props.getString("tone", "neutral")),
        accessibilityLabel = node.props.getString("a11y_label").ifEmpty {
            node.props.getString("label")
        },
        accessibilityHint = node.props.getString("a11y_hint"),
    )

    val isInteractive = false
}

class StatusLabelRendererState(node: NativeUINode) {
    private val nodeId = node.id

    var configuration by mutableStateOf(StatusLabelRendererConfiguration(node))
        private set

    fun serverPublished(tree: NativeUITree): Boolean {
        val node = tree.root.findStatusLabelNode(nodeId) ?: return false
        val published = StatusLabelRendererConfiguration(node)
        if (published == configuration) return false

        configuration = published
        return true
    }
}

object StatusLabelRenderer {
    @Composable
    fun Render(node: NativeUINode, modifier: Modifier) {
        val state = remember(node.id) { StatusLabelRendererState(node) }
        val publishedTree = NativeUIBridge.currentTree.value
        LaunchedEffect(publishedTree) {
            publishedTree?.let(state::serverPublished)
        }

        val tokens = if (isSystemInDarkTheme()) NativeUITheme.dark else NativeUITheme.light
        val configuration = state.configuration

        FirstlightStatusLabelControl(
            label = configuration.label,
            colors = resolveStatusLabelTokenColors(tokens, configuration.tone),
            accessibilityDescription = firstlightStatusLabelDescription(
                label = configuration.label,
                accessibilityLabel = configuration.accessibilityLabel,
                accessibilityHint = configuration.accessibilityHint,
            ),
            modifier = modifier,
        )
    }
}

private fun NativeUINode.findStatusLabelNode(id: Int): NativeUINode? {
    if (this.id == id) return this
    children.forEach { child -> child.findStatusLabelNode(id)?.let { return it } }
    return null
}

File: resources/ios/StatusLabelRenderer.swift
import Combine
import SwiftUI
import UIKit

struct StatusLabelRendererConfiguration: Equatable {
    let label: String
    let tone: StatusLabelTone
    let accessibilityLabel: String
    let accessibilityHint: String

    init(node: NativeUINode) {
        let label = node.props.getString("label")

        self.label = label
        self.tone = StatusLabelTone(
            rawValue: node.props.getString("tone", default: "neutral")
        ) ?? .neutral
        self.accessibilityLabel = node.props.getString("a11y_label").isEmpty
            ? label
            : node.props.getString("a11y_label")
        self.accessibilityHint = node.props.getString("a11y_hint")
    }
}

struct StatusLabelRendererState {
    private(set) var configuration: StatusLabelRendererConfiguration
    private let nodeID: Int

    init(node: NativeUINode) {
        nodeID = node.id
        configuration = StatusLabelRendererConfiguration(node: node)
    }

    @discardableResult
    mutating func serverPublished(_ tree: NativeUITree) -> Bool {
        guard let node = statusLabelNode(id: nodeID, in: tree.root) else {
            return false
        }

        let published = StatusLabelRendererConfiguration(node: node)
        guard published != configuration else { return false }

        configuration = published
        return true
    }
}

struct StatusLabelRenderer: View {
    let node: NativeUINode

    @ObservedObject private var themeStore = NativeUITheme.shared
    @Environment(\.colorScheme) private var colorScheme
    @State private var state: StatusLabelRendererState

    init(node: NativeUINode) {
        self.node = node
        _state = State(initialValue: StatusLabelRendererState(node: node))
    }

    var body: some View {
        let configuration = state.configuration
        let traits = UITraitCollection(
            userInterfaceStyle: colorScheme == .dark ? .dark : .light
        )

        FirstlightStatusLabelControl(
            label: configuration.label,
            tokens: FirstlightStatusLabelTokens.from(
                theme: themeStore.resolve(for: colorScheme),
                tone: configuration.tone,
                traits: traits
            ),
            accessibilityLabel: configuration.accessibilityLabel,
            accessibilityHint: configuration.accessibilityHint
        )
        .onReceive(NativeUIBridge.shared.$currentTree.compactMap { $0 }) { tree in
            state.serverPublished(tree)
        }
    }
}

private func statusLabelNode(id: Int, in node: NativeUINode) -> NativeUINode? {
    if node.id == id { return node }

    for child in node.children {
        if let found = statusLabelNode(id: id, in: child) {
            return found
        }
    }

    return nil
}
